control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go

```diff
@@ -165,6 +166,24 @@ func (r *HostedControlPlaneReconciler) Reconcile(ctx context.Context, req ctrl.R
 		}
 	}
 
+	// Determine if active control plane operator image matches the expected release image payload. If it doesn't
+	// need to no-op until the operator is recreated and at the matching version
+	if _, ok := hostedControlPlane.Annotations[hyperv1.ActiveHypershiftOperatorImage]; !ok {
```

Comment:
As mentioned in slack, this approach of checking inside the controllers is subject to time-of-check to time-of-us races when we are in an active reconciliation while this changes. Can you please explain why you went with this approach overmoving the CPO deployment rollout before the HCP reconciliation and wait for the CPO rollout to finish?

Reply:
It's not subjected to time of check races with leader election:

This is mainly due to the fact now the update of the HostedCluster is atomic in the fact that it updates both the releaseImage and the expectedOperatorImage at a given point in time. 

That allows when a control-plane-operator picks up the update to see both changes at the same time and know if it matches the appropriate version. If not it nops below.

Reply:
I don't want to introduce cross dependencies in the hypershift operator having to do advance introspection into the hostedcluster namespace (that isn't going to scale as this single entity manages up to 440 individual clusters in a single control plane). That's the main reason along with having a clean split of logic.

I guess the potential that you are pointing out is if somehow during the process leader election was lost, transferred to someone else, but yet this component kept running? I'm not sure if that's really something specific to this entity and most controllers would be subject to the same scenario in that case.

Reply:
It is, because an active reconciliation could be running while the version in the HCP gets changed. Leaderelection doesn't change anything about that, it just prevents two copies of the HCO running at the same time.

There is no advanced checking needed, we can find out through the deployment status if the latest version got rolled out (spec.Replics == status.readyReplicas == status.replicas == status.updatedReplicas)

Reply:
at the start of the function here:
[https://github.com/openshift/hypershift/blob/7637a9237e2dccdfbcb35499ab02827ea317b[…]controllers/hostedcontrolplane/hostedcontrolplane_controller.go](https://github.com/openshift/hypershift/blob/7637a9237e2dccdfbcb35499ab02827ea317b178/control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go#L138)

Let’s say after that the HC changes: that doesn’t matter because it’s not going to get picked up in that reconcilation loop


It’s already loaded it, then the next reconciliation loop when it gets the data again: It would see the updated data and see the check at the start and then start noping until a leader elected component had the proper image.

Reply:
So let’s take your example a component in active reconciliation

[Tyler Lisowski](https://app.slack.com/team/U015GA40J4E)  [27 minutes ago](https://coreos.slack.com/archives/C01C8502FMM/p1643992442662849?thread_ts=1643992413.467429&cid=C01C8502FMM)
It has already loaded it’s version of the hcp that it is “reconciling” locally

[Tyler Lisowski](https://app.slack.com/team/U015GA40J4E)  [26 minutes ago](https://coreos.slack.com/archives/C01C8502FMM/p1643992500412159?thread_ts=1643992413.467429&cid=C01C8502FMM)
at the start of the function here:
[https://github.com/openshift/hypershift/blob/7637a9237e2dccdfbcb35499ab02827ea317b[…]controllers/hostedcontrolplane/hostedcontrolplane_controller.go](https://github.com/openshift/hypershift/blob/7637a9237e2dccdfbcb35499ab02827ea317b178/control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go#L138)
Let’s say after that the HC changes: that doesn’t matter because it’s not going to get picked up in that reconcilation loop

[Tyler Lisowski](https://app.slack.com/team/U015GA40J4E)  [25 minutes ago](https://coreos.slack.com/archives/C01C8502FMM/p1643992554015949?thread_ts=1643992413.467429&cid=C01C8502FMM)
It’s already loaded it, then the next reconciliation loop when it gets the data again: It would see the updated data and see the check at the start and then start noping until a leader elected component had the proper image.

[Alvaro Aleman](https://app.slack.com/team/UT4JLAZHQ)  [19 minutes ago](https://coreos.slack.com/archives/C01C8502FMM/p1643992901879319?thread_ts=1643992413.467429&cid=C01C8502FMM)
Okay so everything the CPO could be doing wrong in a version mismatch case also requires the observed HCP to be in a version that differs?

[Tyler Lisowski](https://app.slack.com/team/U015GA40J4E)  [16 minutes ago](https://coreos.slack.com/archives/C01C8502FMM/p1643993060607149?thread_ts=1643992413.467429&cid=C01C8502FMM)
Right yes it all depends on the data it’s seeing locally

[Tyler Lisowski](https://app.slack.com/team/U015GA40J4E)  [16 minutes ago](https://coreos.slack.com/archives/C01C8502FMM/p1643993085497369?thread_ts=1643992413.467429&cid=C01C8502FMM)
This mainly was for my knowledge because I was concerned concurrent reconcile loops could be going on in the same component :slightly_smiling_face:

[Tyler Lisowski](https://app.slack.com/team/U015GA40J4E)  [16 minutes ago](https://coreos.slack.com/archives/C01C8502FMM/p1643993103086459?thread_ts=1643992413.467429&cid=C01C8502FMM)
but I guess even then we would be covered because when it has the data

[Alvaro Aleman](https://app.slack.com/team/UT4JLAZHQ)  [9 minutes ago](https://coreos.slack.com/archives/C01C8502FMM/p1643993490363929?thread_ts=1643992413.467429&cid=C01C8502FMM)
No, the controller machinery only allows one concurrent reconcile per distinct key

[Tyler Lisowski](https://app.slack.com/team/U015GA40J4E)  [< 1 minute ago](https://coreos.slack.com/archives/C01C8502FMM/p1643994068925949?thread_ts=1643992413.467429&cid=C01C8502FMM)
will copy this slack convo to get too just thought it be easier for us to chat here real quick